### PR TITLE
support RFC 31 `hostlist` and `rank` job constraints

### PR DESCRIPTION
**Pull request by grondo**
This PR adds support for `hostlist` and `ranks` constraints operators to librlist and thus sched-simple. This allows requiring or excluding resources based on hostname and/or ranks.

While these operators are now supported in the scheduler, no convenient interface for generating constraints with these operators will be available until #4871 is merged and a follow on PR to use the parser in flux-mini is merged. In that PR we can add end-to-end testing of these new operators in flux-core.

This PR can now be merged independent from #4871.

### CODE REVIEW

**Review by garlick — CHANGES_REQUESTED** [review]
Nicely done.  I called up a couple of unlikely memory leaks on error paths inline which should be corrected.  A couple other minor points:
- space between `*` and function name for functions that return pointers is inconsistent with the rest of the code base
-  casual readers might confuse local function calls prefixed with `idset_` and `hostlist_`  with methods of those classes
- I assume that these functions are not required set/preserve errno on failure.  If the intent was otherwise, then I think I spotted a couple places that need attention and I can call those out.

File: src/common/librlist/match.c [hunk 1]
```diff
@@ -47,6 +48,67 @@ static struct job_constraint *job_constraint_new (flux_error_t *errp)
...
+        if (hostlist_append (hl, json_string_value (entry)) < 0) {
+            char *s = json_dumps (hostlists, 0);
+            errprintf (errp,
+                       "invalid hostlist '%s' in %s",
+                       json_string_value (entry),
+                       s);
+            free (s);
+            goto error;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (grondo) [reply src/common/librlist/match.c:143]: `error:` label has
```
error:
    hostlist_destroy (hl);
    return NULL;
```

though maybe I'm missing something.

File: src/common/librlist/match.c [hunk 2]
```diff
@@ -48,6 +49,77 @@ static struct job_constraint *job_constraint_new (flux_error_t *errp)
...
+        if (add_idset_string (idset, json_string_value (entry)) < 0) {
+            char *s = json_dumps (idsets, 0);
+            errprintf (errp,
+                       "invalid idset '%s' in %s",
+                       json_string_value (entry),
+                       s);
+            free (s);
+            goto error;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (grondo) [reply src/common/librlist/match.c:84]: `error` label has

```c
error:
    idset_destroy (idset);
    return NULL;
```

**Comment by grondo** [comment]
> space between * and function name for functions that return pointers is inconsistent with the rest of the code base

Oops I didn't pay much attention to this one and apparently have been guilty of this before (there are many existing occurrences of this violation in the code, I will fix those later)

>  I assume that these functions are not required set/preserve errno on failure.

No the important functions need to return an error string to job submitters or a job exception and errno would be unhelpful there, so `flux_error_t` is used.

File: src/common/librlist/match.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (garlick) [reply src/common/librlist/match.c:143]: Oh wow, not enough coffee, sorry!

File: src/common/librlist/match.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (garlick) [reply src/common/librlist/match.c:84]: Ibid :coffee:!

**Comment by grondo** [comment]
Ok, I fixed the whitespace, naming, and memory leak errors called out by @garlick. I accidentally squashed the separate commits adding the `hostlist` and `ranks` operators while trying to fix whitespace, but I'll just keep them together in one commit as that seems a little simpler anyway.

**Review by garlick — APPROVED** [review]
Thanks LGTM!

**Comment by grondo** [comment]
Great! Thanks! I'll set MWP.

**Pull request merged**